tests/many_to_many/tests.py

```diff
@@ -165,6 +166,116 @@ def test_slow_add_ignore_conflicts(self):
                 self.a1.publications.add(self.p1)
         mocked.assert_called_once()
 
+    def test_add_relations(self):
+        self.assertIn(self.p1, self.a1.publications.all())
+        self.assertNotIn(self.p3, self.a3.publications.all())
+        self.assertNotIn(self.p4, self.a3.publications.all())
+        self.assertNotIn(self.p3, self.a4.publications.all())
+        self.assertNotIn(self.p4, self.a4.publications.all())
+        with self.assertNumQueries(2):
+            Article.publications.add_relations([
+                (self.a1, self.p1),
+                (self.a3, self.p3),
+                (self.a3, self.p4),
+                (self.a4, self.p3),
+                (self.a4, self.p4),
+            ])
+        self.assertIn(self.p1, self.a1.publications.all())
+        self.assertIn(self.p3, self.a3.publications.all())
+        self.assertIn(self.p4, self.a3.publications.all())
+        self.assertIn(self.p3, self.a4.publications.all())
+        self.assertIn(self.p4, self.a4.publications.all())
+
+    def test_add_relations_by_pk(self):
+        self.assertIn(self.p1, self.a1.publications.all())
+        self.assertNotIn(self.p3, self.a3.publications.all())
+        self.assertNotIn(self.p4, self.a3.publications.all())
+        self.assertNotIn(self.p3, self.a4.publications.all())
+        self.assertNotIn(self.p4, self.a4.publications.all())
+        with self.assertNumQueries(2):
+            Article.publications.add_relations([
+                (self.a1.pk, self.p1.pk),
+                (self.a3.pk, self.p3.pk),
+                (self.a3.pk, self.p4.pk),
+                (self.a4.pk, self.p3.pk),
+                (self.a4.pk, self.p4.pk),
+            ])
+        self.assertIn(self.p1, self.a1.publications.all())
+        self.assertIn(self.p3, self.a3.publications.all())
+        self.assertIn(self.p4, self.a3.publications.all())
+        self.assertIn(self.p3, self.a4.publications.all())
+        self.assertIn(self.p4, self.a4.publications.all())
+
+    def test_add_relations_with_assert_no_collisions(self):
+        # No collision
+        self.assertNotIn(self.p3, self.a3.publications.all())
+        with self.assertNumQueries(1):
+            Article.publications.add_relations([
+                (self.a3, self.p3),
+            ], assert_no_collisions=True)
+        self.assertIn(self.p3, self.a3.publications.all())
+
+        # Collision! Should raise IntegrityError. Should affect no rows.
+        self.assertNotIn(self.p4, self.a4.publications.all())
+        self.assertIn(self.p1, self.a1.publications.all())
+        with transaction.atomic():
+            with self.assertRaisesMessage(IntegrityError, ''):
```

Comment:
I think this should be in a separate test, if possible?

Reply:
Will split test to:

* test_add_relations_with_assert_no_collisions_and_no_collision
* test_add_relations_with_assert_no_collisions_and_collision